core/src/main/java/io/grpc/NameResolver.java

```diff
@@ -120,6 +120,27 @@ public void refresh() {}
      * @since 1.0.0
      */
     public abstract String getDefaultScheme();
+
+    /**
+     * Obtains a name resolver from the factory implementation. If the factory implements
+     * {@link ProxyAwareFactory} then {@link ProxyAwareFactory#newNameResolver(URI, Attributes,
+     * ProxyDetector)} is used instead
+     */
+    public static NameResolver getNameResolver(URI targetUri, Attributes params,
+        ProxyDetector proxyDetector, Factory factory) {
+      if (factory instanceof ProxyAwareFactory) {
+        return ((ProxyAwareFactory)factory).newNameResolver(targetUri, params, proxyDetector);
+      } else {
+        return factory.newNameResolver(targetUri, params);
+      }
+    }
+  }
+
+  public static interface ProxyAwareFactory {
```

Comment:
We typically shy away from interfaces on the public API because they aren't easy to modify later.   This would probably need to become an abstract class, refactored to compose with the nameresolver.

Reply:
I see. My intention was to add the new method to the factory instead of changing the current one, and breaking the API compatibility. Pushed another approach that keeps the same goal without the interface